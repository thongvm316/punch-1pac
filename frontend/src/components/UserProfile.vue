<template>
  <form class="setting-form">
    <div
      class="form-group"
      :class="{ 'has-error': errors.avatar }"
    >
      <label class="form-label">{{ $t('label.avatar') }}</label>
      <img
        class="img-profile"
        :src="targetUser.avatar_url"
        :alt="targetUser.name"
      >
      <input
        class="form-input"
        type="file"
        @change="setAvatarFile"
      >
      <p
        v-if="errors.avatar"
        class="form-input-hint"
      >
        {{ $t('label.avatar') }} {{ errors.avatar[0] }}
      </p>
    </div>
    <div
      class="form-group"
      :class="{ 'has-error': $v.params.email.$anyError || errors.email }"
    >
      <label class="form-label">{{ $t('label.email') }}</label>
      <input
        v-model="$v.params.email.$model"
        class="form-input"
        type="text"
      >
      <p
        v-if="$v.params.email.$anyError && !errors.email"
        class="form-input-hint"
      >
        <span v-if="!$v.params.email.required">{{ $t('validation.required', { name: $t('label.email') }) }}</span>
        <span v-if="!$v.params.email.email">{{ $t('validation.invalid', { name: $t('label.email') }) }}</span>
      </p>
      <p
        v-if="errors.email"
        class="form-input-hint"
      >
        {{ $t('label.email') }} {{ errors.email[0] }}
      </p>
    </div>
    <div
      class="form-group"
      :class="{ 'has-error': $v.params.name.$error || errors.name }"
    >
      <label class="form-label">{{ $t('label.name') }}</label>
      <input
        v-model="$v.params.name.$model"
        class="form-input"
        type="text"
      >
      <p
        v-if="$v.params.name.$error && !errors.name"
        class="form-input-hint"
      >
        {{ $t('validation.required', { name: $t('label.name') }) }}
      </p>
      <p
        v-if="errors.name"
        class="form-input-hint"
      >
        {{ $t('label.name') }} {{ errors.name[0] }}
      </p>
    </div>
    <div class="form-group">
      <label class="form-label">{{ $t('label.gender') }}</label>
      <label class="form-radio">
        <input
          v-model="params.gender"
          type="radio"
          value="male"
        >
        <i class="form-icon" /> {{ $t('meta.gender.male') }}
      </label>
      <label class="form-radio">
        <input
          v-model="params.gender"
          type="radio"
          value="female"
        >
        <i class="form-icon" /> {{ $t('meta.gender.female') }}
      </label>
    </div>
    <div class="form-group">
      <label class="form-label">{{ $t('label.position') }}</label>
      <input
        v-model="params.position"
        class="form-input"
        type="text"
      >
    </div>
    <div class="form-group">
      <label class="form-label">{{ $t('label.role') }}</label>
      <select
        v-model="params.role"
        class="form-select"
        :disabled="!$auth('User', currentUser, targetUser).canEditRole()"
      >
        <option
          v-for="(role, key) in meta.roles"
          :key="key"
          :value="role"
        >
          {{ $t(`meta.roles.${role}`) }}
        </option>
      </select>
    </div>
    <div class="form-group">
      <button
        type="button"
        class="btn btn-success btn-submit"
        :disabled="isDisabled"
        @click="updateUser"
      >
        {{ $t('button.common.save') }}
      </button>
    </div>
  </form>
</template>

<script>
import { mapState, mapMutations } from 'vuex'
import handleSuccess from '../mixins/handle-success'
import userProfileValidate from '../validations/user-profile-validate'
import * as types from '../store/mutation-types'
import axios from 'axios'
import 'formdata-polyfill'

export default {

  mixins: [handleSuccess, userProfileValidate],

  props: {
    targetUser: {
      type: Object,
      required: true
    },
    objectType: {
      type: String,
      default: ''
    }
  },
  data() {
    return {
      params: {
        avatar: '',
        gender: '',
        name: '',
        position: '',
        email: '',
        role: ''
      },
      errors: {}
    }
  },

  computed: {
    ...mapState('initialStates', ['meta'])
  },

  watch: {
    targetUser: function() {
      this.params = { ...this.targetUser }
    }
  },

  created() {
    this.params = { ...this.targetUser }
  },

  methods: {
    ...mapMutations('initialStates', [types.INITIAL_STATES_UPDATE_USER]),

    ...mapMutations('group', [types.UPDATE_GROUP_USER]),

    ...mapMutations('companyUsers', [types.UPDATE_USER]),

    updateUser() {
      let formData = new FormData()
      Object.keys(this.params).forEach(key => formData.set(`user[${key}]`, this.params[key] || ''))

      axios
        .put(`/users/${this.targetUser.id}`, formData, { headers: { 'Content-Type': 'multipart/form-data' } })
        .then(response => {
          if (this.targetUser.id === this.currentUser.id) {
            this[types.INITIAL_STATES_UPDATE_USER](response.data)
          }
          if (this.objectType === 'company') this[types.UPDATE_USER](response.data)
          if (this.objectType === 'group') this[types.UPDATE_GROUP_USER](response.data)
          this.handleSuccess({
            emitType: 'afterUserProfileUpdated',
            message: this.$t('messages.user.updateProfileSuccess')
          })
          this.errors = {}
        })
        .catch(error => {
          if (error.response && error.response.status === 422) this.errors = error.response.data.errors
        })
    },

    setAvatarFile(e) {
      const files = e.target.files || e.dataTransfer.files
      if (!files.length) return
      this.params.avatar = files[0]
    }
  }
}
</script>
