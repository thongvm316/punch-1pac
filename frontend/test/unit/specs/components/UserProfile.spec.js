import { shallowMount } from '@vue/test-utils'
import wrapperOps from '../../supports/wrapper'
import initialStatesData from '../../supports/fixtures/initial-states.api'
import error422 from '../../supports/fixtures/errors.api'
import setComputed from '../../supports/set-computed'
import Repositories from '@/repository'
import UserProfile from '@/components/UserProfile'
jest.mock('@/repository/users')

const targetUser = { ...initialStatesData.currentUser }
const updateUser = jest.spyOn(UserProfile.methods, 'updateUser')
const setAvatarFile = jest.fn()
const handleSuccess = jest.spyOn(UserProfile.mixins[0].methods, 'handleSuccess')
const INITIAL_STATES_UPDATE_USER = jest.spyOn(UserProfile.methods, 'INITIAL_STATES_UPDATE_USER')
const UPDATE_GROUP_USER = jest.spyOn(UserProfile.methods, 'UPDATE_GROUP_USER')
const UPDATE_USER = jest.spyOn(UserProfile.methods, 'UPDATE_USER')

const localWrapperOps = {
  ...wrapperOps,
  methods: {
    setAvatarFile
  },
  propsData: {
    targetUser
  },
  computed: {
    isDisabled() {
      return false
    },
    meta() {
      return initialStatesData.meta
    },
    currentUser() {
      return { ...initialStatesData.currentUser }
    }
  }
}

describe('UserProfile.vue', () => {
  let wrapper

  beforeEach(() => {
    wrapper = shallowMount(UserProfile, localWrapperOps)
  })

  afterEach(() => { wrapper.vm.$destroy() })

  describe('when UserProfile was mounted', () => {
    it('should render correctly', () => {
      expect(wrapper.exists()).toBeTruthy()
      expect(wrapper.isVueInstance()).toBeTruthy()
      expect(wrapper).toMatchSnapshot()
    })
  })

  describe('when methods', () => {
    describe('when setAvatarFile', () => {
      it('should setAvatarFile toHaveBeenCalled', () => {
        wrapper.find('input[type=file]').trigger('change')

        expect(setAvatarFile).toHaveBeenCalled()
      })
    })

    describe('when updateUser', () => {
      const response = { data: initialStatesData.currentUser }
      const mockError = error422

      describe('when targetUser === currentUser', () => {
        it('should INITIAL_STATES_UPDATE_USER', async () => {
          Repositories.get('users').updateUser.mockResolvedValue(response)
          wrapper.find({ ref: 'updateUserButton' }).trigger('click')
          await wrapper.vm.$nextTick()

          expect(updateUser).toHaveBeenCalled()
          expect(INITIAL_STATES_UPDATE_USER).toHaveBeenCalledWith(response.data)
          expect(handleSuccess).toHaveBeenCalledWith({
            emitType: 'afterUserProfileUpdated',
            message: 'Your profile is updated'
          })
          expect(wrapper.vm.errors).toEqual({})
        })

        it('should rejectedError', async () => {
          Repositories.get('users').updateUser.mockRejectedValue(mockError)
          wrapper.find({ ref: 'updateUserButton' }).trigger('click')
          await wrapper.vm.$nextTick()

          expect(updateUser).toHaveBeenCalled()
          expect(wrapper.vm.errors).toEqual(mockError.response.data.errors)
        })
      })

      describe('when objectType === company', () => {
        it('should UPDATE_USER', async () => {
          wrapper.setProps({
            objectType: 'company'
          })
          await wrapper.vm.$nextTick()

          Repositories.get('users').updateUser.mockResolvedValue(response)
          wrapper.find({ ref: 'updateUserButton' }).trigger('click')
          await wrapper.vm.$nextTick()

          expect(UPDATE_USER).toHaveBeenCalledWith(response.data)
        })
      })

      describe('when objectType === group', () => {
        it('should UPDATE_USER', async () => {
          wrapper.setProps({
            objectType: 'group'
          })
          await wrapper.vm.$nextTick()

          Repositories.get('users').updateUser.mockResolvedValue(response)
          wrapper.find({ ref: 'updateUserButton' }).trigger('click')
          await wrapper.vm.$nextTick()

          expect(UPDATE_GROUP_USER).toHaveBeenCalledWith(response.data)
        })
      })
    })
  })
})
