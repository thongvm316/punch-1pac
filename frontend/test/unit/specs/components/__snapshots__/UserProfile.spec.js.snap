// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserProfile.vue when UserProfile was mounted should render correctly 1`] = `
<form class="setting-form">
  <div class="form-group"><label class="form-label">Avatar</label> <img src="http://localhost:3000/static/default/user_avatar-751fbd8e7c012f38bb132002fc9d7fab3461c512847921987c2fab06e376b263.png" alt="John Kenedy" class="img-profile"> <input type="file" class="form-input">
    <!---->
  </div>
  <div class="form-group"><label class="form-label">Email</label> <input type="text" class="form-input">
    <!---->
    <!---->
  </div>
  <div class="form-group"><label class="form-label">Name</label> <input type="text" class="form-input">
    <!---->
    <!---->
  </div>
  <div class="form-group"><label class="form-label">Gender</label> <label class="form-radio"><input type="radio" value="male"> <i class="form-icon"></i> Male
    </label> <label class="form-radio"><input type="radio" value="female"> <i class="form-icon"></i> Female
    </label></div>
  <div class="form-group"><label class="form-label">Position</label> <input type="text" class="form-input"></div>
  <div class="form-group"><label class="form-label">Role</label> <select disabled="disabled" class="form-select">
      <option value="member">
        Member
      </option>
      <option value="admin">
        Admin
      </option>
      <option value="superadmin">
        Superadmin
      </option>
    </select></div>
  <div class="form-group"><button type="button" class="btn btn-success btn-submit">
      Save
    </button></div>
</form>
`;
